irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -174,37 +149,63 @@ namespace iroha {
 
       auto top_block_height = getTopBlockHeight();
       if (not top_block_height) {
-        log_->error("Unable to continue without knowing top block height");
+        log_->error("Failed to synchronize: could not get my top block height.");
         return;
       }
 
-      int64_t height_diff = msg.round.block_round - *top_block_height;
+      const shared_model::interface::types::HeightType required_height =
+          SynchronizationOutcomeType::kCommit == alternative_outcome
+          ? msg.round.block_round
+          : msg.round.block_round - 1;
+      int64_t height_diff = required_height - *top_block_height;
+
       if (height_diff < 0) {
         log_->info(
             "Storage is already in synchronized state. Top block height is {}",
             *top_block_height);
         return;
       }
 
-      if (SynchronizationOutcomeType::kCommit != alternative_outcome
-          and (0 == height_diff or 1 == height_diff)) {
+      if (height_diff == 0) {
         notifier_.get_subscriber().on_next(
             // TODO: nickaleks IR-147 18.01.19 add peers
             // list from GateObject when it has one
             SynchronizationEvent{
-                rxcpp::observable<>::empty<
-                    std::shared_ptr<shared_model::interface::Block>>(),
                 alternative_outcome,
                 msg.round,
                 nullptr /* this is going to be fixed in PR-2066 (IR-147) */});
         return;
       }
 
-      auto result =
-          downloadMissingBlocks(msg, *top_block_height, alternative_outcome);
-      if (result) {
-        notifier_.get_subscriber().on_next(*result);
+      assert(height_diff > 0);
+
+      auto ledger_state = downloadMissingBlocks(
+          *top_block_height, required_height, msg.public_keys);
+
+      shared_model::interface::types::HeightType new_height;
+      if (ledger_state) {
+        new_height = (*ledger_state)->height;
+      } else {
+        const auto opt_new_height = getTopBlockHeight();
```

Comment:
It might be simpler to assume that a missing ledger state is a fatal error at the moment.